pkg/customers/customer_search.go

```diff
@@ -84,54 +96,103 @@ func (r *sqlCustomerRepository) searchCustomers(params SearchParams) ([]*client.
 	}
 	defer stmt.Close()
 
-	var customerIDs []string
 	customers := make([]*client.Customer, 0)
-
-	// grab customerIDs
 	rows, err := stmt.Query(args...)
 	if err != nil {
-		return customers, err
+		return nil, err
 	}
 	for rows.Next() {
-		var customerID string
-		if err := rows.Scan(&customerID); err == nil {
-			customerIDs = append(customerIDs, customerID)
-		} else {
-			return customers, err
+		var c client.Customer
+		var birthDate *string
+		err := rows.Scan(
+			&c.CustomerID,
+			&c.FirstName,
+			&c.MiddleName,
+			&c.LastName,
+			&c.NickName,
+			&c.Suffix,
+			&c.Type,
+			&birthDate,
+			&c.Status,
+			&c.Email,
+			&c.CreatedAt,
+			&c.LastModified,
+		)
+		if err != nil {
+			return nil, err
 		}
+		customers = append(customers, &c)
 	}
 
-	// Read each Customer
-	for i := range customerIDs {
-		cust, err := r.GetCustomer(customerIDs[i])
-		if err != nil {
-			return customers, fmt.Errorf("search: customerID=%s error=%v", customerIDs[i], err)
-		}
-		customers = append(customers, cust)
+	if len(customers) == 0 {
+		return customers, nil
+	}
+
+	var customerIDs []string
+	for _, c := range customers {
+		customerIDs = append(customerIDs, c.CustomerID)
+	}
+
+	phonesByCustomerID, err := r.getPhones(customerIDs)
+	if err != nil {
+		return nil, fmt.Errorf("fetching customer phones: %v", err)
+	}
+	addressesByCustomerID, err := r.getAddresses(customerIDs)
+	if err != nil {
+		return nil, fmt.Errorf("fetching customer addresses: %v", err)
+	}
+	metadataByCustomerID, err := r.getMetadata(customerIDs)
+	if err != nil {
+		return nil, fmt.Errorf("fetching customer metadata: %v", err)
 	}
+
+	for _, c := range customers {
+		c.Phones = phonesByCustomerID[c.CustomerID]
+		c.Addresses = addressesByCustomerID[c.CustomerID]
+		c.Metadata = metadataByCustomerID[c.CustomerID].Metadata
+	}
+
 	return customers, nil
 }
 
 func buildSearchQuery(params SearchParams) (string, []interface{}) {
 	var args []interface{}
-	query := `select customer_id from customers where deleted_at is null and organization = ?`
-	args = append(args, params.Organization)
+	query := `select customer_id, first_name, middle_name, last_name, nick_name, suffix, type, birth_date, status, email, created_at, last_modified
+from customers where deleted_at is null`
+
+	if params.Organization != "" {
+		query += " and organization = ?"
+		args = append(args, params.Organization)
+	}
+
 	if params.Query != "" {
-		query += " and lower(first_name) || \" \" || lower(last_name) LIKE ?"
-		args = append(args, "%"+params.Query+"%")
+		// warning: this will ONLY work for MySQL
```

Comment:
Unfortunately, I haven't found a way to add this clause such that both sqlite and mysql will be satisfied

Reply:
Maybe it's a sign to switch to only one DB? :)